src/internal/packager2/deploy.go

```diff
@@ -0,0 +1,704 @@
+	Retries int
+	// Number of layers to push concurrently per image
+	OCIConcurrency int
+	// Whether to use plainHTTP when connecting to the registry
+	PlainHTTP bool
+	// Whether or not to skipTLSVerify when connecting to the registry
+	InsecureTLSSkipVerify bool
+	// How to configure Zarf state if it's not already been configured
+	GitServer      types.GitServerInfo
+	RegistryInfo   types.RegistryInfo
+	ArtifactServer types.ArtifactServerInfo
+
+	// [Library Only] A map of component names to chart names containing Helm Chart values to override values on deploy
+	ValuesOverridesMap map[string]map[string]map[string]interface{}
+}
+
+// deployer tracks mutable fields across deployments. Because components can create a cluster and create state
+// any of these fields are subject to change from one component to the next
+type deployer struct {
+	s           *state.State
+	c           *cluster.Cluster
+	vc          *variables.VariableConfig
+	hpaModified bool
+}
+
+func Deploy(ctx context.Context, pkgLayout *layout.PackageLayout, opts DeployOpts) ([]types.DeployedComponent, error) {
+	l := logger.From(ctx)
+	l.Info("starting deploy", "package", pkgLayout.Pkg.Metadata.Name)
+	start := time.Now()
+	variableConfig, err := getPopulatedVariableConfig(ctx, pkgLayout.Pkg, opts.SetVariables)
+	if err != nil {
+		return nil, err
+	}
+
+	d := deployer{
+		vc: variableConfig,
+	}
+
+	// During deploy we disable
+	defer d.resetRegistryHPA(ctx)
+	l.Debug("variables populated", "time", time.Since(start))
+
+	deployedComponents, err := d.deployComponents(ctx, pkgLayout, opts)
+	if err != nil {
+		return nil, err
+	}
+	if len(deployedComponents) == 0 {
+		l.Warn("no components were selected for deployment. Inspect the package to view the available components and select components interactively or by name with \"--components\"")
+	}
+	l.Debug("deployment complete", "duration", time.Since(start))
+	return deployedComponents, nil
+}
+
+func (d *deployer) resetRegistryHPA(ctx context.Context) {
+	l := logger.From(ctx)
+	if d.c != nil && d.hpaModified {
+		if err := d.c.EnableRegHPAScaleDown(ctx); err != nil {
+			l.Debug("unable to reenable the registry HPA scale down", "error", err.Error())
+		}
+	}
+}
+
+func (d *deployer) isConnectedToCluster() bool {
+	return d.c != nil
+}
+
+func (d *deployer) deployComponents(ctx context.Context, pkgLayout *layout.PackageLayout, opts DeployOpts) ([]types.DeployedComponent, error) {
+	l := logger.From(ctx)
+	deployedComponents := []types.DeployedComponent{}
+
+	for _, component := range pkgLayout.Pkg.Components {
+		packageGeneration := 1
+		// Connect to cluster if a component requires it.
+		if component.RequiresCluster() {
+			timeout := cluster.DefaultTimeout
+			if pkgLayout.Pkg.IsInitConfig() {
+				timeout = 5 * time.Minute
+			}
+			connectCtx, cancel := context.WithTimeout(ctx, timeout)
+			defer cancel()
+			if !d.isConnectedToCluster() {
+				var err error
+				d.c, err = cluster.NewWithWait(ctx)
```

Comment:
Should this be utilizing `connectCtx`? 
```suggestion
				d.c, err = cluster.NewWithWait(connectCtx)
```

Reply:
Yes, really great catch